### PR TITLE
Feature/df 3138 avp page one backend frontend sync

### PR DESCRIPTION
**Pull request by mdavidovski**
Changes:

- Using @OneToMany instead of @ElementCollection for AvpDocumentUploads. This way we take care of the update/delete issue, and we don't lose the attachments.

- Adding an attachment type validation when creating attachments. Meaning if we add to an employment entry an attachment type which is not for employments, we will throw an exception.

- Other small changes (validations, nullability handling, code cleanup, naming)

### CODE REVIEW

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java [hunk 1]
```diff
@@ -341,6 +335,7 @@ public AvpDocumentUploadDto createAttachment(AvpId id, AvpDocumentUploadId avpDo
         Avp avp = getAvp(id);
         AvpDocumentUpload avpDocumentUpload = avp.findAvpDocumentUpload(avpDocumentUploadId)
                 .orElseThrow(() -> new IllegalStateException("Missing AvpDocumentUpload " + avpDocumentUploadId + " for creating Attachment in Avp with ID : " + id + "."));
+        validateAttachmentsForDocumentUploadType(avpDocumentUpload.getType(), inputStream.getType());
```
Comment (emina-str) [inline application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java:338]: inputStream is just very bad naming :) I almost wrote you to close it :D :D 
could you please just rename the parameter?

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/AvpDocumentUpload.java [hunk 2]
```diff
@@ -8,33 +8,39 @@
...
         // For reflexion libs
     }
 
-    public AvpDocumentUpload(AvpDocumentUploadId id, AvpDocumentUploadType type, EntryId entryId, boolean required) {
+    public AvpDocumentUpload(AvpDocumentUploadId id, AvpDocumentUploadType type, EntryId entryId, boolean required, AvpId avpId) {
         this.id = Condition.notNull(id);
         this.type = Condition.notNull(type);
         this.entryId = entryId;
         this.required = required;
+        this.avpId = avpId;
```
Comment (emina-str) [inline domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/AvpDocumentUpload.java:43]: you're missing a getter here

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/Avp.java [hunk 3]
```diff
@@ -385,7 +382,6 @@ public void updateSelfEmployments(Boolean hasSelfEmployment, Set<SelfEmploymentE
     }
 
     private void updateHasSelfEmployment(Boolean hasSelfEmployment) {
-        checkNotClosed();
```
Comment (emina-str) [inline domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/Avp.java:388]: why did you remove "checkNotClosed()" ?

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java [hunk 4]
```diff
@@ -509,20 +562,69 @@ public void testUpdateEmployments() {
         assertThat(returnedEmploymentsDto.getEmployments().iterator().next().getToDate()).isNull();
         assertThat(returnedEmploymentsDto.getEmployments().iterator().next().getCertificateSender()).isNull();
         assertThat(returnedEmploymentsDto.getEmployments().iterator().next().getCompany()).isNull();
+        // Required documents
```
Comment (emina-str) [inline application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:565]: can you just use the Avp Document Uploads or something ... so it's not misleading

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java [hunk 5]
```diff
@@ -540,11 +642,42 @@ public void testRemoveEmployments() {
 
         // then repository
         Set<EmploymentEntry> employments = this.avpRepository.findAllEmploymentEntryByAvpId(AVP01.id());
+        Set<AvpDocumentUpload> avpDocumentUploads = this.avpRepository.findAllDocumentUploadsByAvpId(AVP01.id());
         assertThat(employments).isEmpty();
+        assertThat(avpDocumentUploads.stream().filter(it -> it.getType() == AvpDocumentUploadType.EMPLOYMENT)).isEmpty();
```
Comment (emina-str) [inline application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:647]: maybe also check are the AvpDocumentUpload empty at this point.

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java [hunk 6]
```diff
@@ -509,20 +562,69 @@ public void testUpdateEmployments() {
...
         assertThat(employments.iterator().next().getToDate()).isNull();
         assertThat(employments.iterator().next().getCertificateSender()).isNull();
         assertThat(employments.iterator().next().getCompany()).isNull();
+        // Required documents
+        assertThat(avpDocumentUploads.stream().filter(it -> it.getType() == AvpDocumentUploadType.EMPLOYMENT)).isNotEmpty().hasSize(1);
+        assertThat(avpDocumentUploads.stream().filter(it -> it.getType() == AvpDocumentUploadType.EMPLOYMENT).iterator().next().getEntryId()).isEqualTo(id);
+    }
+
+    @Test
+    public void testUpdateEmploymentsMultiple() {
```
Comment (emina-str) [inline application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:586]: We had a failing case on the frontend ... with f.e. 3 Employments, of which maybe 2 had the CertificateSender.SELF, and you try to delete on of the Employment with CertificateSender.SELF and then it deleted both of them.
(because of the sql checking also required field when deleting). maybe to combine this test to add additional Employment and then after updating (and checking if all is executing fine) on the second step to delete one of them and check again do we have all the correct stuff left.

Do you understand what I'm saying?

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java [hunk 7]
```diff
@@ -42,10 +39,7 @@
 
 import java.time.LocalDateTime;
 import java.time.YearMonth;
-import java.util.List;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
```
Comment (palsch) [inline application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java:42]: please dont use wildcard for imports (you can adjust that in Intellij properties)

**Review by palsch — APPROVED** [review]
see my comments

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java (hunk 7 above)
Comment (mdavidovski) [reply application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java:42]: We consider that bad practice? I thought if there are more than a few, we add the wildcard?

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java (hunk 1 above)
Comment (mdavidovski) [reply application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java:338]: Yeah, I missed that one...

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/AvpDocumentUpload.java (hunk 2 above)
Comment (mdavidovski) [reply domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/AvpDocumentUpload.java:43]: We don't need it. The avpId is only added so JPA recognizes the mapping from the Avp entity, where we specify 
@OneToMany(mappedBy = "avpId"...

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java (hunk 7 above)
Comment (palsch) [reply application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java:42]: I think, it is not a big deal to add them separately using Intellij and it can optimize/solve some possible issues that could come up by using wildcards.
See also https://medium.com/@tharakamd.12/is-it-bad-to-use-wildcard-imports-in-java-1b46a863b2be

File: domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/Avp.java (hunk 3 above)
Comment (mdavidovski) [reply domain/src/main/java/ch/admin/seco/alv/service/onlineform/domain/avp/Avp.java:388]: Both here and in the updateHasEmployment I removed it because they are duplicates. We already do this check in the public method, or updateSelfEmployments()

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java (hunk 4 above)
Comment (mdavidovski) [reply application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:565]: Yeah, I missed that one

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java (hunk 5 above)
Comment (mdavidovski) [reply application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:647]: What do you mean?

Comment (mdavidovski) [reply application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:647]: Maybe you mean the attachments?

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java (hunk 6 above)
Comment (mdavidovski) [reply application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:586]: Yeah, but that test case is covered further down in `testRemoveOneEmployment()`.
But I'll refactor some stuff just in case

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java (hunk 7 above)
Comment (mdavidovski) [reply application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java:42]: The default import number before adding a wildcard is 5. 
To which number should I set it?

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java (hunk 5 above)
Comment (emina-str) [reply application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:647]: sorry, my bad, I saw something else when I wrote it :D

File: application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java (hunk 6 above)
Comment (emina-str) [reply application/src/test/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationServiceTest.java:586]: i saw it later on ... yeah, it's not maybe necessary to add also those changes. it's ok for me!

File: application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java (hunk 7 above)
Comment (palsch) [reply application/src/main/java/ch/admin/seco/alv/service/onlineform/application/avp/AvpApplicationService.java:42]: I set it to 99 in my Intellij

**Review by emina-str — APPROVED** [review]
